core/src/main/java/org/apache/calcite/plan/Digest.java

```diff
@@ -0,0 +1,256 @@
+      if (item.right instanceof RelNode) {
+        RelNode input = (RelNode) item.right;
+        normalized.add(Pair.of(item.left, simpleRelDigest(input)));
+      } else {
+        normalized.add(item);
+      }
+    }
+    return normalized;
+  }
+
+  /**
+   * Returns a simple string format digest.
+   *
+   * <p>Currently, returns composition of class name and id.
+   *
+   * @param rel The rel
+   */
+  private static String simpleRelDigest(RelNode rel) {
+    return rel.getRelTypeName() + '#' + rel.getId();
+  }
+
+  @Override public String toString() {
+    if (null != digest) {
+      return digest;
+    }
+    StringBuilder sb = new StringBuilder();
+    sb.append(rel.getRelTypeName());
+
+    for (RelTrait trait : rel.getTraitSet()) {
+      sb.append('.');
+      sb.append(trait.toString());
+    }
+
+    sb.append('(');
+    int j = 0;
+    for (Pair<String, Object> item : items) {
+      if (j++ > 0) {
+        sb.append(',');
+      }
+      sb.append(item.left);
+      sb.append('=');
+      sb.append(item.right);
+    }
+    sb.append(')');
+    digest = sb.toString();
+    return digest;
+  }
+
+  @Override public int compareTo(@Nonnull Digest other) {
+    return this.equals(other) ? 0 : this.rel.getId() - other.rel.getId();
+  }
+
+  @Override public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    Digest that = (Digest) o;
+    return hashCode == that.hashCode && deepEquals(that);
+  }
+
+  /**
+   * The method is used to resolve hash conflict, in current 6000+ tests, there are about 8
+   * tests with conflict, so we do not cache the hash code items in order to
+   * reduce mem consumption.
+   */
+  private boolean deepEquals(Digest other) {
+    Object[] thisItems = collect(this.rel, this.items, true);
+    Object[] thatItems = collect(other.rel, other.items, true);
+    if (thisItems.length != thatItems.length) {
+      return false;
+    }
+    for (int i = 0; i < thisItems.length; i++) {
+      if (!Objects.equals(thisItems[i], thatItems[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override public int hashCode() {
+    return hashCode;
+  }
+
+  /**
+   * Creates a digest with given rel and properties.
+   */
+  public static Digest create(RelNode rel, List<Pair<String, Object>> contents) {
+    return new Digest(rel, contents);
+  }
+
+  /**
+   * Creates a digest with given rel.
+   */
+  public static Digest create(RelNode rel) {
+    return new Digest(rel);
+  }
+
+  /**
+   * Creates a digest with given rel and string format digest
+   */
+  public static Digest create(RelNode rel, String digest) {
+    return new Digest(rel, digest);
+  }
+
+  /**
+   * Instantiates a digest with solid string format digest, this digest should only
+   * be used as a initial.
+   */
+  public static Digest initial(RelNode rel) {
+    return new Digest(rel, simpleRelDigest(rel));
+  }
+}
```

Comment:
What's the difference between `Digest#create` and `Digest#initial` ? Their results seem the same.

Reply:
To distinguish that the digest is just as a initial which should only be used once in the constructor.

Reply:
It's really hard to distinguish these two methods. Is it possible to remove `Digest#create` which seems useless?

Reply:
I can remove the `#initial` and we use the `#create` uniformly.